ArC blogpost (#256)
quarkusio/quarkusio.github.io: _posts/2019-07-24-arc.adoc

---
layout: post
date:   2019-07-24 00:00 +0100
author: mkouba
synopsis: Quarkus ArC is a build-time oriented dependency injection based on CDI 2.0. But what does it actually mean and what benefits does a build-time processing DI bring?
---

= Quarkus Dependency Injection

Quarkus ArC is a build-time oriented dependency injection based on http://cdi-spec.org/[CDI 2.0].
In this blogpost, we're going to explain the relationship to the specification and describe some of the benefits and drawbacks of the build-time processing design.

== Compatibility

It does not make sense to reinvent the wheel, especially when it comes to dependency injection.

> Threre are many frameworks that try to solve similar problems.

Reviewer: ```suggestion
There are many frameworks that try to solve similar problems.
```

A year ago, we made a design decision to build Quarkus DI on top of CDI.
We had some very good reasons to choose CDI:

1. CDI is a mature and proven component model
2. We at Red Hat have almost ten years of experience with developing http://weld.cdi-spec.org/[Weld - the CDI Reference Implementation]

> 3. CDI API is built on top of `javax.inject` so that it should be easy to migrate from any DI framework compatible with `@Inject`

Reviewer: ```suggestion
3. The CDI API is built on top of `javax.inject` so that it should be easy to migrate from any DI framework compatible with `@Inject`
```
maybe?

> Our primary goal was to implement supersonic build-time oriented DI solution compatible with CDI.

Reviewer: ```suggestion
Our primary goal was to implement supersonic a build-time oriented DI solution compatible with CDI.
```
?

Author: well, it should be probably `a supersonic build-time oriented DI` instead of `supersonic a build-time oriented DI`

Reviewer: Ah yes, sorry :)

This would allow users to continue using CDI in their applications but also leverage Quarkus build-time optimizations.
However, ArC is not a full CDI implementation verified by the TCK - see also the list of https://quarkus.io/guides/cdi-reference#supported_features[supported features] and the list of https://quarkus.io/guides/cdi-reference#limitations[limitations].

== Build-time Processing Benefits and Drawbacks

=== Fail Fast

Beans and dependencies are validated during the build so that your application can never fail in production with common problems such as `AmbiguousResolutionException` or `UnsatisfiedResolutionException`.

=== Instant Startup

When the app starts ArC just loads all the metadata and initializes some internal structures.
There is no need to analyze the application classes again.
This means negligible startup overhead.

> NOTE: This applies to both - the GraalVM and OpenJDK HotSpot runtimes.

Reviewer: ```suggestion
NOTE: This applies to both the GraalVM and OpenJDK HotSpot runtimes.
```

=== Minimized Runtime

In Quarkus 0.19, ArC plus integration runtime consist of 72 classes and occupies ~ 140 KB in jars.
http://weld.cdi-spec.org/[Weld 3.1.1] (CDI Reference Implementation) core is roughly 1200 classes and approx. 2 MB jar.
In other words, ArC runtime takes approx. 7% of the Weld runtime in terms of number of classes and jar footprint.

=== Optimized Code Paths and Metadata Structures

When generating the metadata classes ArC can sometimes produce leaner and smarter logic because it has the whole application analyzed already.
This is one of the areas where we would like to develop and improve ArC.

=== Extension Points

Unfortunately, CDI portable extensions are inherently runtime constructs and therefore cannot be fully supported in Quarkus.
In fact, all CDI extensions are ignored at the moment.
Nevertheless, most of the functionality can be achieved using https://quarkus.io/guides/cdi-reference#build-time-extension-points[Quarkus extensions].

> We have an idea how to support simple stateless CDI extensions in the future but it's not a matter of high priority.

Author: Hm, I don't find this sentence "unduly negative", it just does not have a high priority. But I'm going to remove the sentence completely because TBH I don't think we should implement this ;-).

=== Qualified Injected Fields

Normally, if you declare an injected field you'll always need to use `@Inject` and optionally required qualifiers:
 
[source,java]
----
  @Inject
  @ConfigProperty(name = "cool")
  String coolProperty;
----

In Quarkus, you can skip the `@Inject` annotation completely if an injected field declares at least one qualifier:

[source,java]
----
  @ConfigProperty(name = "cool")
  String coolProperty;
----

NOTE: `@Inject` is still required for constructor and method injection.

=== Simplified Constructor Injection

In CDI, a normal scoped bean must always declare a no-args constructor (this constructor is normally generated by the compiler unless you declare any other constructor).
However, this requirement complicates constructor injection - you would need to provide a dummy no-args constructor to make things work in CDI.

[source,java]
----
@ApplicationScoped
public class MyCoolService {

  private SimpleProcessor processor;

  MyCoolService() { // dummy constructor needed
  }
  
  @Inject // constructor injection
  MyCoolService(SimpleProcessor proessor) {
    this.processor = processor;
  }
}
----

> There is no need to declare dummy constructors for normal scoped bean in Quarkus - it's generated automatically.

Reviewer: ```suggestion
There is no need to declare dummy constructors for normal scoped bean in Quarkus - they are generated automatically.
```
?

> Also if there's only one constructor there is no need for @Inject.

Reviewer: ```suggestion
Also if there's only one constructor there is no need for `@Inject`.
```

In this case, a bean that can be overriden should be annotated with `@io.quarkus.arc.DefaultBean`.
And that's it.

[source,java]
----
@DefaultBean
class MyBean {
}

class MyOwnBean extends MyBean {
}

@Inject // MyOwnBean wins and is injected!
MyBean bean;
----

=== Removing Unused Beans

GraalVM native image does a pretty good job when removing all classes that are not reachable from your application.
However, sometimes checking reachability is not enough.
Sometimes the framework itself must decide whether a component is needed or not.
In standard CDI, all beans are retained by the container no matter whether they're needed or not.

Let's suppose we have a bean class `org.acme.Foo`.
This bean class imports and uses a lot of various classes.
It's annotated with `@ApplicationScoped` and so Quarkus needs to generate a bean metadata class and a client proxy and register this metadata when the application starts.
But what if nobody ever uses this bean?
We would still hold a reference on the generated metadata, and the bean class itself, and its dependencies. 
In other words, all these classes would be reachable.

Quarkus attempts to remove all *unused beans* during build by default.

> This helps to reduce the number of generated classes and also the amout of memory needed at runtime.

Reviewer: ```suggestion
This helps to reduce the number of generated classes and also the amount of memory needed at runtime.
```
